### ISSUE TITLE
Pool with async commands?

### ISSUE THREAD

**Issue by allen-servedio** [lerouxrgd/clj-lettuce#7]
Does the new pool functionality work with async commands? In looking at the macro, it appears that the connection is potentially returned to the pool before the async commands would have finished.

**Comment by lerouxrgd** [comment]
Would this solve your use-case ?

```clj
(defmacro with-conn-pool-async
  "Like with-conn-pool but takes an extra promise-done argument and
  will wait on it before returning the connection to the pool"
  [conn-pool cmds-name promise-done & body]
  `(let [conn# (borrow-conn ~conn-pool)
         ~cmds-name ((:cmds-fn ~conn-pool)
                     (assoc (:connector ~conn-pool) :stateful-conn conn#))]
     (try
       ~@body
       (finally @~promise-done ;; wait for completion
                (return-conn ~conn-pool conn#)))))
```

Where as an end user you define when the async work is done (and the connection is then returned to the pool).
Like in this example:

```clj
(testing "async pooled transaction"
    (let [conn-pool (pool/conn-pool redis-conn conn/commands-sync)
          p-res (promise)]
      (pool/with-conn-pool-async conn-pool cmds p-res
        (future ;; simulating async commands
          (redis/multi cmds)
          (redis/set   cmds :c 3)
          (redis/set   cmds :d 4)
          (redis/get   cmds :c)
          (redis/get   cmds :d)
          (is (= ["OK" "OK" 3 4] (redis/exec cmds)))
          (deliver p-res :done)))
      (is (= :done @p-res))
      (pool/close conn-pool)))
```

**Comment by allen-servedio** [comment]
I think that is certainly one approach, but doesn't that block the thread until the async process completes?

Since you have a manifold module, maybe it makes sense to only support this functionality within that context? That way you could create a deferred for them that they could send a success or error to. Additionally, you could alter your try-finally to be a try-catch and return the connection to the pool should the body throw an exception.

It could be that people would like both approaches - so I still think what you show here has merit.

**Comment by lerouxrgd** [comment]
How about this ?

```clj
(defmacro with-conn-pool*
  "Like with-conn-pool but also binds the pooled connection to `conn-name`.
  User is responsible for returning it to the pool"
  [conn-pool cmds-name conn-name & body]
  `(let [~conn-name (borrow-conn ~conn-pool)
         ~cmds-name ((:cmds-fn ~conn-pool)
                     (assoc (:connector ~conn-pool) :stateful-conn ~conn-name))]
     ~@body))
```

And with `celtuce.manifold` you could use it like this:

```clj
(testing "async pooled transaction with manifold"
    (let [conn-pool (pool/conn-pool redis-conn commands-manifold)
          res (pool/with-conn-pool* conn-pool cmds c
                (d/chain (redis/multi cmds)
                         (redis/set   cmds :c 3)
                         (redis/set   cmds :d 4)
                         (redis/get   cmds :c)
                         (redis/get   cmds :d)                 
                         (fn [_] (redis/exec cmds))
                         (fn [res]
                           (pool/return-conn conn-pool c)
                           res)) )]
      (is (= ["OK" "OK" 3 4] @res))
      (pool/close conn-pool)))
```

**Comment by allen-servedio** [comment]
Looks good to me - seems like a reasonable approach.

**Comment by lerouxrgd** [comment]
Released in v0.2.2

**Comment by allen-servedio** [comment]
Looks good @lerouxrgd. You may want to add an example on https://github.com/lerouxrgd/celtuce/blob/master/modules/celtuce-pool/README.md so that folks are aware of how to use it.
